[WIP] MCH: new digit time definition (obsolete)

aferrero2707 opened the pull request:
The new definition combines the SAMPA time stamp, the SAMPA bunch crossing counter,
and the orbit number from the HBFrame containing the digit.
    
The introduction of this new time definition requires adding the bunch crossing counter to the
SampaCluster strcture, and therefore to change the Elink decoders as well as some of the unity tests.
    
The new definition does not change the size of the Digit structure.

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
[MCH] implemented new definition of digit time

The new definition combines the SAMPA time stamp, the SAMPA bunch crossing counter,
and the orbit number of the HBFrame containing the digit.

The introduction of this new time definition requires adding the bunch crossing counter to the
SampaCluster strcture, and therefore to change the Elink decoders as well as some of the unity tests.

The new definition does not change the size of the Digit structure.

Commit:
[MCH] added two reserved bits in the hit time structure

Commit:
[MCH] fixed bug in custom mapping loader code

aphecetche commented on Detectors/MUON/MCH/Base/include/MCHBase/HitTime.h:
```diff
@@ -0,0 +1,41 @@
+namespace mch
+{
+
+struct HitTime {
```
The first comment I'd make is about naming : I don't like HitTime as I feel Hit is really too much connoted "simulation". Given that struct is only used (at least in this PR) in Digit, why not make it a Time struct of the Digit itself ?

aferrero2707 replied:
Time definition moved inside Digit structure, as you proposed.

aphecetche commented on Detectors/MUON/MCH/Base/src/Digit.cxx:
```diff
@@ -29,7 +35,7 @@ bool Digit::operator==(const Digit& other) const
          mPadID == other.mPadID &&
          mADC == other.mADC &&
-         closeEnough(mTime, other.mTime);
+         closeEnough(mTime.sampaTime, other.mTime.sampaTime);
```
with a time that's now an int, closeEnough is no longer needed

aferrero2707 replied:
Ok, removed.

aphecetche commented on Detectors/MUON/MCH/Base/src/Digit.cxx:
```diff
@@ -19,7 +19,13 @@ bool closeEnough(double x, double y, double eps = 1E-6)
 }
 
-Digit::Digit(double time, int detid, int pad, unsigned long adc)
+Digit::Digit(int sampaTime, int detid, int pad, unsigned long adc)
```
while we're at changing the Digit ctor, I would go a step further and move the time as the last parameter (and with a default value of 0 ?)

aferrero2707 replied:
I have moved the time parameter at the end, and removed the other constructor that takes a double value.
As a consequence, I had to modify quite some files in the encoder part, and it would be good if you could have a closer look.

aphecetche commented on Detectors/MUON/MCH/Raw/Common/include/MCHRawCommon/SampaCluster.h:
```diff
@@ -43,15 +43,15 @@ struct SampaCluster {
   /// if some parameter does not fit within its expected range
   /// a std::invalid_argument exception is thrown.
-  explicit SampaCluster(uint16_t timestamp, uint32_t chargeSum);
+  explicit SampaCluster(uint16_t timestamp, uint32_t bunchCrossing, uint32_t chargeSum);
```
maybe for consistency we should rename the first parameter (and corresponding data member) to "sampaTime" ?

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Raw/Common/include/MCHRawCommon/SampaCluster.h:
```diff
@@ -43,15 +43,15 @@ struct SampaCluster {
   /// if some parameter does not fit within its expected range
   /// a std::invalid_argument exception is thrown.
-  SampaCluster(uint16_t timestamp, const std::vector<uint16_t>& samples);
+  SampaCluster(uint16_t timestamp, uint32_t bunchCrossing, const std::vector<uint16_t>& samples);
```
also for clarity I was thinking to use uint10_t and uint20_t typedefs to make the API even clearer. What do you think ?

aferrero2707 replied:
Good idea, I introduced the two types in the SampaCluster structure, via a "using" keyword.

aphecetche commented on Detectors/MUON/MCH/Raw/Common/src/SampaCluster.cxx:
```diff
@@ -21,16 +21,18 @@ namespace mch
-  : timestamp(impl::assertIsInRange("timestamp", timestamp, 0, 0x3FF)),   // 10 bits
+SampaCluster::SampaCluster(uint16_t timestamp, uint32_t bunchCrossing, uint32_t chargeSum)
+  : timestamp(impl::assertIsInRange("timestamp", timestamp, 0, 0x3FF)),
+    bunchCrossing(impl::assertIsInRange("bunchCrossing", bunchCrossing, 0, 0xFFFFF)),
```
if the bunchcrossing is only 20 bits, that should be asserted here

aferrero2707 replied:
I though it would be asserted by
```
impl::assertIsInRange("bunchCrossing", bunchCrossing, 0, 0xFFFFF)
```
Did I miss something?

aphecetche commented on Detectors/MUON/MCH/Raw/Common/test/testSampaCluster.cxx:
```diff
@@ -26,36 +26,37 @@ BOOST_AUTO_TEST_SUITE(o2_mch_raw)
 BOOST_AUTO_TEST_SUITE(sampacluster)
 
 uint16_t defaultTimestamp{0x3FF};
+uint32_t defaultBunchCrossing{0xFFFF};
```
that's not a valid default is BC is 20 bits, right ? 
should also add a test to actually ensure the SampaCluster is throwing if an attempt is made to construct it with more than 20-bits BC.

aferrero2707 replied:
I fixed it. I still need to see how to add a test for the BC range.

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/src/testUserLogicEndpointDecoder.cxx:
```diff
@@ -241,7 +241,7 @@ BOOST_AUTO_TEST_CASE(SampleModeSimplest)
   // only one channel with one very small cluster
   // fitting within one 64-bits word
-  SampaCluster cl(345, {123, 456});
+  SampaCluster cl(345, 6789, {123, 456});
```
shouldn't we use the newly added BC information to the string constructed for each cluster ?

aferrero2707 replied:
I guess so, but I am not yet much familiar with the tests, so I need to have a detailed look.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -152,8 +160,9 @@ class DataDecoderTask
+      content += " ";
     }
-    return s;
+    return content;
```
given the number of bugs I've introduced in this class, we should really think about proving a unit test for it ;-)

aferrero2707 replied:
Actually, I expect the custom mapping to be useful only in the early stages of the development and testing, and only for very specific cases, so a unit test might be over-doing...

aphecetche reviewed (commented):
@aferrero2707 the conflict that is there now is because of the merging of #3589 a few minutes ago. There are more changes in the works (moving to RDHUtils as much as possible) that affects the encoders, so I'd propose to come back with possibly more comments to this one a bit later on, hopefully tomorrow.

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
Merge branch 'dev' into mch-new-digit-time

aphecetche commented:
@aferrero2707 FYI the work I was referring to is now in the check pipeline #3622

Commit:
[MCH] moved digit time structure inside Digit class definition

The time parameter has been also moved to the last position in the Digit constructor,
and the old "double" parameter has been removed.

This modification required several changes in the Raw encoder and simulation parts.

Commit:
[MCH] renamed timestamp as sampaTime in SampaCluster struct

Commit:
[MCH] fixed default bunch crossing value in SampaCluster testing code

Commit:
Merge remote-tracking branch 'aphecetche/mch-new-digit-time' into mch-new-digit-time

aferrero2707 commented:
This PR was incorporated into #3660 and merged into dev.
